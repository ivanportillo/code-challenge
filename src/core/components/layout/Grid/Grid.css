.Grid {
    display: flex;
    flex-direction: column;
    flex: 1;
    align-items: center;
}

.Row {
    display: flex;
    width: 100%;
}
