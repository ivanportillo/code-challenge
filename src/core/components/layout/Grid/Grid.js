import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import './Grid.css';

const Column = ({ children, gutter, columnWidth }) => {
  const ColumnStyle = {
    width: `${columnWidth}%`,
    marginRight: gutter / 2,
    marginLeft: gutter / 2,
  };
  return (
    <div style={ColumnStyle}>
      {children}
    </div>
  );
};

const Grid = ({ children, columns, gutter, spaceBetween }) => {
  const rows = _.chunk(children, columns);
  const rowStyle = {
    marginBottom: spaceBetween,
  };
  return (
    <div className="Grid">
      {rows.map(chunk =>
        <div key={chunk[0].key} style={rowStyle} className="Row">
          {chunk.map(child =>
            <Column key={child.key} gutter={gutter} columnWidth={100 / columns}>
              {child}
            </Column>,
          )}
        </div>,
      )}
    </div>
  );
};

Grid.propTypes = {
  children: PropTypes.node.isRequired,
  columns: PropTypes.number.isRequired,
  gutter: PropTypes.number.isRequired,
  spaceBetween: PropTypes.number.isRequired,
};

export default Grid;
