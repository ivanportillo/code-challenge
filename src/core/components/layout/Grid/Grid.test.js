import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import Grid from './Grid';

describe('Grid', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <Grid columns={3} gutter={2} spaceBetween={2}>
        <div key="1">Element 1</div>
        <div key="2">Element 2</div>
        <div key="3">Element 3</div>
        <div key="4">Element 1</div>
        <div key="5">Element 2</div>
        <div key="6">Element 3</div>
      </Grid>, div);
  });

  it('should render a grid', () => {
    const grid = shallow(
      <Grid columns={3} gutter={2} spaceBetween={2}>
        <div key="1">Element 1</div>
        <div key="2">Element 2</div>
        <div key="3">Element 3</div>
        <div key="4">Element 1</div>
        <div key="5">Element 2</div>
        <div key="6">Element 3</div>
      </Grid>);

    expect(grid).toMatchSnapshot();
  });
});
