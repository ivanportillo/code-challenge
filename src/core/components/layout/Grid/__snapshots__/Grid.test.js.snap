exports[`Grid should render a grid 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="Grid">
    <div
        className="Row"
        style={
            Object {
                "marginBottom": 2,
              }
        }>
        <Column
            columnWidth={33.333333333333336}
            gutter={2}>
            <div>
                Element 1
            </div>
        </Column>
        <Column
            columnWidth={33.333333333333336}
            gutter={2}>
            <div>
                Element 2
            </div>
        </Column>
        <Column
            columnWidth={33.333333333333336}
            gutter={2}>
            <div>
                Element 3
            </div>
        </Column>
    </div>
    <div
        className="Row"
        style={
            Object {
                "marginBottom": 2,
              }
        }>
        <Column
            columnWidth={33.333333333333336}
            gutter={2}>
            <div>
                Element 1
            </div>
        </Column>
        <Column
            columnWidth={33.333333333333336}
            gutter={2}>
            <div>
                Element 2
            </div>
        </Column>
        <Column
            columnWidth={33.333333333333336}
            gutter={2}>
            <div>
                Element 3
            </div>
        </Column>
    </div>
</div>,
  "nodes": Array [
    <div
      className="Grid">
      <div
            className="Row"
            style={
                  Object {
                        "marginBottom": 2,
                      }
            }>
            <Column
                  columnWidth={33.333333333333336}
                  gutter={2}>
                  <div>
                        Element 1
                  </div>
            </Column>
            <Column
                  columnWidth={33.333333333333336}
                  gutter={2}>
                  <div>
                        Element 2
                  </div>
            </Column>
            <Column
                  columnWidth={33.333333333333336}
                  gutter={2}>
                  <div>
                        Element 3
                  </div>
            </Column>
      </div>
      <div
            className="Row"
            style={
                  Object {
                        "marginBottom": 2,
                      }
            }>
            <Column
                  columnWidth={33.333333333333336}
                  gutter={2}>
                  <div>
                        Element 1
                  </div>
            </Column>
            <Column
                  columnWidth={33.333333333333336}
                  gutter={2}>
                  <div>
                        Element 2
                  </div>
            </Column>
            <Column
                  columnWidth={33.333333333333336}
                  gutter={2}>
                  <div>
                        Element 3
                  </div>
            </Column>
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Grid
        columns={3}
        gutter={2}
        spaceBetween={2}>
        <div>
                Element 1
        </div>
        <div>
                Element 2
        </div>
        <div>
                Element 3
        </div>
        <div>
                Element 1
        </div>
        <div>
                Element 2
        </div>
        <div>
                Element 3
        </div>
</Grid>,
      "_debugID": 23,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "children": Array [
            <div>
              Element 1
</div>,
            <div>
              Element 2
</div>,
            <div>
              Element 3
</div>,
            <div>
              Element 1
</div>,
            <div>
              Element 2
</div>,
            <div>
              Element 3
</div>,
          ],
          "columns": 3,
          "gutter": 2,
          "spaceBetween": 2,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="Grid">
          <div
                    className="Row"
                    style={
                              Object {
                                        "marginBottom": 2,
                                      }
                    }>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 1
                              </div>
                    </Column>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 2
                              </div>
                    </Column>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 3
                              </div>
                    </Column>
          </div>
          <div
                    className="Row"
                    style={
                              Object {
                                        "marginBottom": 2,
                                      }
                    }>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 1
                              </div>
                    </Column>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 2
                              </div>
                    </Column>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 3
                              </div>
                    </Column>
          </div>
</div>,
        "_debugID": 24,
        "_renderedOutput": <div
          className="Grid">
          <div
                    className="Row"
                    style={
                              Object {
                                        "marginBottom": 2,
                                      }
                    }>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 1
                              </div>
                    </Column>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 2
                              </div>
                    </Column>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 3
                              </div>
                    </Column>
          </div>
          <div
                    className="Row"
                    style={
                              Object {
                                        "marginBottom": 2,
                                      }
                    }>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 1
                              </div>
                    </Column>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 2
                              </div>
                    </Column>
                    <Column
                              columnWidth={33.333333333333336}
                              gutter={2}>
                              <div>
                                        Element 3
                              </div>
                    </Column>
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Grid
    columns={3}
    gutter={2}
    spaceBetween={2}>
    <div>
        Element 1
    </div>
    <div>
        Element 2
    </div>
    <div>
        Element 3
    </div>
    <div>
        Element 1
    </div>
    <div>
        Element 2
    </div>
    <div>
        Element 3
    </div>
</Grid>,
}
`;
